Tests/OpenCombineTests/PublisherTests/MapTests.swift

```diff
@@ -0,0 +1,156 @@
+//
+//  MapTests.swift
+//
+//
+//  Created by Anton Nazarov on 25/06/2019.
+//
+
+import XCTest
+
+#if OPENCOMBINE_COMPATIBILITY_TEST
+import Combine
+#else
+import OpenCombine
+#endif
+
+@available(macOS 10.15, *)
+final class MapTests: XCTestCase {
+    static let allTests = [
+        ("testEmpty", testEmpty),
+        ("testError", testError),
+        ("testRange", testRange),
+        ("testDemandSubscribe", testDemandSubscribe),
+        ("testDemandSend", testDemandSend),
+        ("testCompletion", testCompletion),
+        ("testCancel", testCancel),
+        ("testCancelAlreadyCancelled", testCancelAlreadyCancelled)
+    ]
+
+    func testEmpty() {
+        // Given
+        let tracking = TrackingSubscriberBase<Never>(receiveSubscription: { $0.request(.unlimited) })
+        let publisher = PassthroughSubject<Int, Never>()
+        // When
+        publisher.map { $0 * 2 }.subscribe(tracking)
+        // Then
+        XCTAssertEqual(tracking.subscriptions.count, 1)
+        XCTAssertEqual(tracking.history, [
+            .subscription(Subscriptions.empty)
+        ])
+    }
+
+    func testError() {
+        // Given
+        let expectedError = TestingError.oops
+        let tracking = TrackingSubscriber(receiveSubscription: { $0.request(.unlimited) })
+        let publisher = PassthroughSubject<Int, TestingError>()
+        // When
+        publisher.map { $0 * 2 }.subscribe(tracking)
+        publisher.send(completion: .failure(expectedError))
+        // Then
+        XCTAssertEqual(tracking.subscriptions.count, 1)
+        XCTAssertEqual(tracking.history, [
+            .subscription(Subscriptions.empty),
+            .completion(Subscribers.Completion<TestingError>.failure(expectedError))
+        ])
+    }
+
+    func testRange() {
+        // Given
+        let publisher = PassthroughSubject<Int, TestingError>()
+        let map = publisher.map { $0 * 2 }
+        let tracking = TrackingSubscriber(receiveSubscription: { $0.request(.unlimited) })
+        // When
+        publisher.send(1)
+        map.subscribe(tracking)
+        publisher.send(2)
+        publisher.send(3)
+        publisher.send(completion: .finished)
+        publisher.send(5)
+        // Then
+        XCTAssertEqual(tracking.history, [
+            .subscription(Subscriptions.empty),
+            .value(4),
+            .value(6),
+            .completion(.finished)
+        ])
+    }
+
+    func testDemandSubscribe() {
+        // Given
+        let expectedSubscribeDemand = 42
+        let subscription = CustomSubscription()
+        let publisher = CustomPublisher(subscription: subscription)
+        let map = publisher.map { $0 * 2 }
+        let tracking = TrackingSubscriber(receiveSubscription: { $0.request(.max(expectedSubscribeDemand)) })
```

Comment:
Here you test that if a downstream subscriber (`tracking`) requests some elements in the `receiveSubscription` callback, the same number of elements is requested from the upstream publisher.

This is useful, but let's also make sure that if the downstream subscriber doesn't request any demand at all, the upstream publisher won't receive any requests.

Reply:
Implemented additional test case